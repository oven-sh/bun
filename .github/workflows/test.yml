name: Test

concurrency:
  group: ${{ github.workflow }}-${{ inputs.run-id || github.run_id }}
  cancel-in-progress: true

on:
  push:
  workflow_call:
    inputs:
      run-id:
        type: string
        required: true

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/download-artifact@v4
        with:
          run-id: ${{ inputs.run-id || '8623134412' }}
          pattern: bun-*
      - run: |
          ls -la
