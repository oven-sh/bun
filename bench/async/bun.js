import { bench, run } from "mitata";

bench("sync", () => {});
bench("async", async () => {});
bench("await 1", async () => await 1);

await run();
