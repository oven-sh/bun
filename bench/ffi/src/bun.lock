{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@napi-rs/cli": "^2.10.1",
        "@node-rs/helper": "^1.3.3",
      },
    },
  },
  "packages": {
    "@napi-rs/cli": ["@napi-rs/cli@2.14.8", "", { "bin": { "napi": "scripts/index.js" } }, "sha512-IvA3s8BqohMdUbOkFn7+23u1dhIJZCkA8Xps7DD4SLdCMbcbUF6MUuKiqxuqmVHBFTaxU25sU56WdX3efqGgPw=="],

    "@napi-rs/triples": ["@napi-rs/triples@1.1.0", "", {}, "sha512-XQr74QaLeMiqhStEhLn1im9EOMnkypp7MZOwQhGzqp2Weu5eQJbpPxWxixxlYRKWPOmJjsk6qYfYH9kq43yc2w=="],

    "@node-rs/helper": ["@node-rs/helper@1.3.3", "", { "dependencies": { "@napi-rs/triples": "^1.1.0" } }, "sha512-p4OdfQObGN9YFy5WZaGwlPYICQSe7xZYyXB0sxREmvj1HzGKp5bPg2PlfgfMZEfnjIA882B9ZrnagYzZihIwjA=="],
  }
}
