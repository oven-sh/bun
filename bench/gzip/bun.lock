{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@babel/standalone": "7.24.10",
      },
    },
  },
  "packages": {
    "@babel/standalone": ["@babel/standalone@7.24.10", "", {}, "sha512-nGC37EKfmelpyCXto1pw6SBkD5ZQRdMbL6WISi28xWit9dtiy9dChU1WgEfzturUTxrmOGkMDRrCydFMA7uOaQ=="],
  }
}
