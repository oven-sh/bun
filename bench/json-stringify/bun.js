import { bench, run } from "mitata";

bench("JSON.stringify({hello: 'world'})", () =>
  JSON.stringify({ hello: "world" }),
);

const otherUint8Array = new Uint8Array(1024);
bench("Uint8Array.from(otherUint8Array)", () =>
  Uint8Array.from(otherUint8Array),
);

run();
