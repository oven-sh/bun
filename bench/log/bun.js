import { bench, run } from "mitata";

bench("console.log('hello')", () => console.log("hello"));
bench("console.log({ hello: 'object' })", () => console.log({ hello: "object" }));
await run();
