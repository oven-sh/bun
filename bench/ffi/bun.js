import { ptr, dlopen, CString, toBuffer } from "bun:ffi";
import { run, bench, group } from "mitata";

const { napiNoop, napiHash, napiString } = require(import.meta.dir + "/src/ffi_napi_bench.node");

const {
  symbols: {
    ffi_noop: { native: ffi_noop },
    ffi_hash: { native: ffi_hash },
    ffi_string: { native: ffi_string },
  },
} = dlopen(import.meta.dir + "/src/ffi_napi_bench.node", {
  ffi_noop: { args: [], returns: "void" },
  ffi_string: { args: [], returns: "ptr" },
  ffi_hash: { args: ["ptr", "u32"], returns: "u32" },
});

const bytes = new Uint8Array(64);

group("bun:ffi", () => {
  bench("noop", () => ffi_noop());
  bench("hash", () => ffi_hash(ptr(bytes), bytes.byteLength));

  bench("c string", () => new CString(ffi_string()));
});

if (process.env.SHOW_NAPI)
  group("bun:napi", () => {
    bench("noop", () => napiNoop());
    bench("hash", () => napiHash(bytes));

    bench("string", () => napiString());
  });

await run();
