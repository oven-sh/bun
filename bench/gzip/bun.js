import { run, bench, group } from "mitata";
import { gzipSync, gunzipSync } from "bun";

const data = new TextEncoder().encode("Hello World!".repeat(9999));

const compressed = gzipSync(data);

const libraries = ["zlib"];
if (Bun.semver.satisfies(Bun.version, ">=1.1.21")) {
  libraries.push("libdeflate");
}

const options = { library: undefined };
const benchFn = (name, fn) => {
  if (libraries.length > 1) {
    group(name, () => {
      for (const library of libraries) {
        bench(library, () => {
          options.library = library;
          fn();
        });
      }
    });
  } else {
    bench(name, () => {
      fn();
    });
  }
};

benchFn(`roundtrip - "Hello World!".repeat(9999))`, () => {
  gunzipSync(gzipSync(data, options), options);
});

benchFn(`gzipSync("Hello World!".repeat(9999)))`, () => {
  gzipSync(data, options);
});

benchFn(`gunzipSync("Hello World!".repeat(9999)))`, () => {
  gunzipSync(compressed, options);
});

await run();
