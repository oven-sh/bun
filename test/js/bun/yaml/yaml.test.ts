import { describe, expect, test } from "bun:test";

describe("Bun.YAML", () => {
  describe("parse", () => {
    test("parses null values", () => {
      expect(Bun.YAML.parse("null")).toBe(null);
      expect(Bun.YAML.parse("~")).toBe(null);
      expect(Bun.YAML.parse("")).toBe(null);
    });

    test("parses boolean values", () => {
      expect(Bun.YAML.parse("true")).toBe(true);
      expect(Bun.YAML.parse("false")).toBe(false);
      expect(Bun.YAML.parse("yes")).toBe(true);
      expect(Bun.YAML.parse("no")).toBe(false);
      expect(Bun.YAML.parse("on")).toBe(true);
      expect(Bun.YAML.parse("off")).toBe(false);
    });

    test("parses number values", () => {
      expect(Bun.YAML.parse("42")).toBe(42);
      expect(Bun.YAML.parse("3.14")).toBe(3.14);
      expect(Bun.YAML.parse("-17")).toBe(-17);
      expect(Bun.YAML.parse("0")).toBe(0);
      expect(Bun.YAML.parse(".inf")).toBe(Infinity);
      expect(Bun.YAML.parse("-.inf")).toBe(-Infinity);
      expect(Bun.YAML.parse(".nan")).toBeNaN();
    });

    test("parses string values", () => {
      expect(Bun.YAML.parse('"hello world"')).toBe("hello world");
      expect(Bun.YAML.parse("'single quoted'")).toBe("single quoted");
      expect(Bun.YAML.parse("unquoted string")).toBe("unquoted string");
      expect(Bun.YAML.parse('key: "value with spaces"')).toEqual({
        key: "value with spaces",
      });
    });

    test("parses arrays", () => {
      expect(Bun.YAML.parse("[1, 2, 3]")).toEqual([1, 2, 3]);
      expect(Bun.YAML.parse("- 1\n- 2\n- 3")).toEqual([1, 2, 3]);
      expect(Bun.YAML.parse("- a\n- b\n- c")).toEqual(["a", "b", "c"]);
      expect(Bun.YAML.parse("[]")).toEqual([]);
    });

    test("parses objects", () => {
      expect(Bun.YAML.parse("{a: 1, b: 2}")).toEqual({ a: 1, b: 2 });
      expect(Bun.YAML.parse("a: 1\nb: 2")).toEqual({ a: 1, b: 2 });
      expect(Bun.YAML.parse("{}")).toEqual({});
      expect(Bun.YAML.parse('name: "John"\nage: 30')).toEqual({
        name: "John",
        age: 30,
      });
    });

    test("parses nested structures", () => {
      const yaml = `
users:
  - name: Alice
    age: 30
    hobbies:
      - reading
      - hiking
  - name: Bob
    age: 25
    hobbies:
      - gaming
      - cooking
`;
      expect(Bun.YAML.parse(yaml)).toEqual({
        users: [
          {
            name: "Alice",
            age: 30,
            hobbies: ["reading", "hiking"],
          },
          {
            name: "Bob",
            age: 25,
            hobbies: ["gaming", "cooking"],
          },
        ],
      });
    });

    test("parses complex nested objects", () => {
      const yaml = `
database:
  host: localhost
  port: 5432
  credentials:
    username: admin
    password: secret
  options:
    ssl: true
    timeout: 30
`;
      expect(Bun.YAML.parse(yaml)).toEqual({
        database: {
          host: "localhost",
          port: 5432,
          credentials: {
            username: "admin",
            password: "secret",
          },
          options: {
            ssl: true,
            timeout: 30,
          },
        },
      });
    });

    test.todo("handles circular references with anchors and aliases", () => {
      const yaml = `
parent: &ref
  name: parent
  child:
    name: child
    parent: *ref
`;
      const result = Bun.YAML.parse(yaml);
      expect(result.parent.name).toBe("parent");
      expect(result.parent.child.name).toBe("child");
      expect(result.parent.child.parent).toBe(result.parent);
    });

    test("handles multiple documents", () => {
      const yaml = `
---
document: 1
---
document: 2
`;
      expect(Bun.YAML.parse(yaml)).toEqual([{ document: 1 }, { document: 2 }]);
    });

    test("handles multiline strings", () => {
      const yaml = `
literal: |
  This is a
  multiline
  string
folded: >
  This is also
  a multiline
  string
`;
      expect(Bun.YAML.parse(yaml)).toEqual({
        literal: "This is a\nmultiline\nstring\n",
        folded: "This is also a multiline string\n",
      });
    });

    test("handles special keys", () => {
      const yaml = `
"special-key": value1
'another.key': value2
123: numeric-key
`;
      expect(Bun.YAML.parse(yaml)).toEqual({
        "special-key": "value1",
        "another.key": "value2",
        "123": "numeric-key",
      });
    });

    test("handles empty values", () => {
      const yaml = `
empty_string: ""
empty_array: []
empty_object: {}
null_value: null
`;
      expect(Bun.YAML.parse(yaml)).toEqual({
        empty_string: "",
        empty_array: [],
        empty_object: {},
        null_value: null,
      });
    });

    test("throws on invalid YAML", () => {
      expect(() => Bun.YAML.parse("[ invalid")).toThrow();
      expect(() => Bun.YAML.parse("{ key: value")).toThrow();
      expect(() => Bun.YAML.parse(":\n :  - invalid")).toThrow();
    });

    test("handles dates and timestamps", () => {
      const yaml = `
date: 2024-01-15
timestamp: 2024-01-15T10:30:00Z
`;
      const result = Bun.YAML.parse(yaml);
      // Dates might be parsed as strings or Date objects depending on implementation
      expect(result.date).toBeDefined();
      expect(result.timestamp).toBeDefined();
    });

    test("preserves object identity for aliases", () => {
      const yaml = `
definitions:
  - &user1
    id: 1
    name: Alice
  - &user2
    id: 2
    name: Bob
assignments:
  project1:
    - *user1
    - *user2
  project2:
    - *user2
`;
      const result = Bun.YAML.parse(yaml);
      expect(result.assignments.project1[0]).toBe(result.definitions[0]);
      expect(result.assignments.project1[1]).toBe(result.definitions[1]);
      expect(result.assignments.project2[0]).toBe(result.definitions[1]);
    });

    test("handles comments", () => {
      const yaml = `
# This is a comment
key: value # inline comment
# Another comment
another: value
`;
      expect(Bun.YAML.parse(yaml)).toEqual({
        key: "value",
        another: "value",
      });
    });

    test("handles flow style mixed with block style", () => {
      const yaml = `
array: [1, 2, 3]
object: {a: 1, b: 2}
mixed:
  - {name: Alice, age: 30}
  - {name: Bob, age: 25}
block:
  key1: value1
  key2: value2
`;
      expect(Bun.YAML.parse(yaml)).toEqual({
        array: [1, 2, 3],
        object: { a: 1, b: 2 },
        mixed: [
          { name: "Alice", age: 30 },
          { name: "Bob", age: 25 },
        ],
        block: {
          key1: "value1",
          key2: "value2",
        },
      });
    });

    test("handles quoted strings with special characters", () => {
      const yaml = `
single: 'This is a ''quoted'' string'
double: "Line 1\\nLine 2\\tTabbed"
unicode: "\\u0041\\u0042\\u0043"
`;
      expect(Bun.YAML.parse(yaml)).toEqual({
        single: "This is a 'quoted' string",
        double: "Line 1\nLine 2\tTabbed",
        unicode: "ABC",
      });
    });

    test("handles large numbers", () => {
      const yaml = `
int: 9007199254740991
float: 1.7976931348623157e+308
hex: 0xFF
octal: 0o777
binary: 0b1010
`;
      const result = Bun.YAML.parse(yaml);
      expect(result.int).toBe(9007199254740991);
      expect(result.float).toBe(1.7976931348623157e308);
      expect(result.hex).toBe(255);
      expect(result.octal).toBe(511);
      expect(result.binary).toBe("0b1010");
    });

    test("handles explicit typing", () => {
      const yaml = `
explicit_string: !!str 123
explicit_int: !!int "456"
explicit_float: !!float "3.14"
explicit_bool: !!bool "yes"
explicit_null: !!null "anything"
`;
      expect(Bun.YAML.parse(yaml)).toEqual({
        explicit_string: "123",
        explicit_int: "456",
        explicit_float: "3.14",
        explicit_bool: "yes",
        explicit_null: "anything",
      });
    });

    test("handles strings that look like numbers", () => {
      const yaml = `
shasum1: 1e18495d9d7f6b41135e5ee828ef538dc94f9be4
shasum2: 19f3afed71c8ee421de3892615197b57bd0f2c8f
`;
      expect(Bun.YAML.parse(yaml)).toEqual({
        shasum1: "1e18495d9d7f6b41135e5ee828ef538dc94f9be4",
        shasum2: "19f3afed71c8ee421de3892615197b57bd0f2c8f",
      });
    });

    test("handles merge keys", () => {
      const yaml = `
defaults: &defaults
  adapter: postgres
  host: localhost
development:
  <<: *defaults
  database: dev_db
production:
  <<: *defaults
  database: prod_db
  host: prod.example.com
`;
      expect(Bun.YAML.parse(yaml)).toEqual({
        defaults: {
          adapter: "postgres",
          host: "localhost",
        },
        development: {
          adapter: "postgres",
          host: "localhost",
          database: "dev_db",
        },
        production: {
          adapter: "postgres",
          host: "prod.example.com",
          database: "prod_db",
        },
      });
    });
  });
});
