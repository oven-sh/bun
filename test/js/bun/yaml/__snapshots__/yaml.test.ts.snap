// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Bun.YAML parse issue 22286 2`] = `
{
  "A Hat in Time": {
    "ActPlando": {
      "Dead Bird Studio Basement": "The Big Parade",
    },
    "ActRandomizer": "insanity",
    "BabyTrapWeight": 0,
    "BadgeSellerMaxItems": 8,
    "BadgeSellerMinItems": 5,
    "BaseballBat": true,
    "CTRLogic": {
      "nothing": 0,
      "scooter": 1,
      "sprint": 0,
      "time_stop_only": 0,
    },
    "ChapterCostIncrement": 5,
    "ChapterCostMinDifference": 5,
    "CompassBadgeMode": "closest",
    "DWAutoCompleteBonuses": true,
    "DWEnableBonus": false,
    "DWExcludeAnnoyingBonuses": true,
    "DWExcludeAnnoyingContracts": true,
    "DWExcludeCandles": true,
    "DWShuffle": false,
    "DWShuffleCountMax": 25,
    "DWShuffleCountMin": 18,
    "DWTimePieceRequirement": 15,
    "DeathWishOnly": false,
    "EnableDLC1": false,
    "EnableDLC2": true,
    "EnableDeathWish": false,
    "EndGoal": {
      "finale": 1,
      "rush_hour": 0,
      "seal_the_deal": 0,
    },
    "ExcludeTour": false,
    "FinalChapterMaxCost": 35,
    "FinalChapterMinCost": 30,
    "FinaleShuffle": false,
    "HatItems": true,
    "HighestChapterCost": 25,
    "LaserTrapWeight": 0,
    "LogicDifficulty": "moderate",
    "LowestChapterCost": 5,
    "MaxExtraTimePieces": "random-range-high-5-8",
    "MaxPonCost": 80,
    "MetroMaxPonCost": 50,
    "MetroMinPonCost": 10,
    "MinExtraYarn": "random-range-middle-5-15",
    "MinPonCost": 20,
    "NoPaintingSkips": true,
    "NoTicketSkips": "rush_hour",
    "NyakuzaThugMaxShopItems": 4,
    "NyakuzaThugMinShopItems": 1,
    "ParadeTrapWeight": 0,
    "RandomizeHatOrder": "time_stop_last",
    "ShipShapeCustomTaskGoal": 0,
    "ShuffleActContracts": true,
    "ShuffleAlpineZiplines": true,
    "ShuffleStorybookPages": true,
    "ShuffleSubconPaintings": true,
    "StartWithCompassBadge": true,
    "StartingChapter": {
      "1": 1,
      "2": 1,
      "3": 1,
    },
    "Tasksanity": false,
    "TasksanityCheckCount": 18,
    "TasksanityTaskStep": 1,
    "TimePieceBalancePercent": "random-range-low-20-35",
    "TrapChance": 0,
    "UmbrellaLogic": true,
    "YarnAvailable": "random-range-middle-40-50",
    "YarnBalancePercent": 25,
    "YarnCostMax": 8,
    "YarnCostMin": 5,
    "accessibility": {
      "full": 1,
      "minimal": 0,
    },
    "death_link": false,
    "exclude_locations": [
      "Queen Vanessa's Manor - Bedroom Chest",
      "Queen Vanessa's Manor - Hall Chest",
      "Act Completion (The Big Parade)",
    ],
    "non_local_items": [
      "Hookshot Badge",
      "Umbrella",
      "Dweller Mask",
    ],
    "priority_locations": [
      "Act Completion (Award Ceremony)",
      "Badge Seller - Item 1",
      "Badge Seller - Item 2",
      "Mafia Boss Shop Item",
      "Bluefin Tunnel Thug - Item 1",
      "Green Clean Station Thug A - Item 1",
      "Green Clean Station Thug B - Item 1",
      "Main Station Thug A - Item 1",
      "Main Station Thug B - Item 1",
      "Main Station Thug C - Item 1",
      "Pink Paw Station Thug - Item 1",
      "Yellow Overpass Thug A - Item 1",
      "Yellow Overpass Thug B - Item 1",
      "Yellow Overpass Thug C - Item 1",
    ],
    "progression_balancing": "random-range-middle-40-50",
    "start_inventory_from_pool": {
      "Sprint Hat": 1,
    },
  },
  "game": "A Hat in Time",
  "name": "niyrme-AHiT{NUMBER}",
  "requires": {
    "version": "0.6.2",
  },
  "x-options-async": {
    "A Hat in Time": {
      "+non_local_items": [
        "Brewing Hat",
        "Ice Hat",
      ],
      "ChapterCostIncrement": 7,
      "ChapterCostMinDifference": 7,
      "EndGoal": {
        "finale": 9,
        "rush_hour": 1,
        "seal_the_deal": 0,
      },
      "FinalChapterMaxCost": 50,
      "FinalChapterMinCost": 40,
      "HighestChapterCost": 40,
      "LowestChapterCost": 10,
      "NoPaintingSkips": false,
      "death_link": false,
      "priority_locations": [],
      "progression_balancing": "random-range-low-10-30",
    },
  },
  "x-options-sync": {
    "A Hat in Time": {
      "+start_inventory_from_pool": {
        "Badge Pin": 1,
      },
      "+triggers": [
        {
          "option_category": "A Hat in Time",
          "option_name": "EndGoal",
          "option_result": "finale",
          "options": {
            "A Hat in Time": {
              "EnableDLC2": false,
              "FinalChapterMaxCost": 35,
              "FinalChapterMinCost": 25,
              "MaxPonCost": 100,
              "MinPonCost": 30,
            },
          },
        },
      ],
    },
  },
}
`;
